### DISCUSSION TITLE
New feature : flag keyword

### DISCUSSION THREAD

**Discussion by Leogiciel** [dotnet/csharplang#367] [General]
Like the enum keyword, but with flag values by default.

Actually, this enum:
```
[Flags]
public enum AccessRights
{
    None,
    Read,
    Write,
    Save,
    Validate
}
```

Assign value 3 to Save, which isn't the desired behaviour for a flag.

This one:
```
[Flags]
    public enum AccessRights
    {
        None = 0,
        Read = 1,
        Write = 2,
        Save = 4,
        Validate = 8
    }
```
...does the trick, but it is low-level coding because developpers should compute every powers of two which seems "old school", especially for long enums.

Would love to see a "flag" keyword encapsulating the bitfield logic:
```
    public flag AccessRights
    {
        None, //0
        Read, //1
        Write, //2
        Save, //4
        Validate //8
    }
```

Cheers :)

**Comment by DavidArno** [comment]
An alternative, to save "working out" powers of 2, would be to use binary literals:
```cs
[Flags]
public enum AccessRights
{
    None = 0,
    Read = 0b0001,
    Write = 0b0010,
    Save = 0b0100,
    Validate = 0b1000
}
```

**Comment by nil4** [comment]
I find using shift operators leads to more readable (and easily reviewable) code:

```csharp
[Flags]
public enum AccessRights
{
    None =     0,
    Read =     1 << 0,
    Write =    1 << 1,
    Save =     1 << 2,
    Validate = 1 << 3
}
```

**Comment by Leogiciel** [comment]
@nil4 method is good, but what if you need to use an intermediate value (eg: ReadWrite) ? Then int values are more readable again... This is off-topic, and goes in the sense that if there is no need to have intermediate values, there is no need either to specify values for a high-level flag.

**Comment by FredyFerrari** [comment]
@Leogiciel what about `ReadWrite = Read | Write,`?

**Comment by Leogiciel** [comment]
I totally agree @FredyFerrari.

I was just arguing about the unneeded technical consideration of assigning values to a bitfield...

**Comment by FredyFerrari** [comment]
@Leogiciel but even your example would be a bit tricky: AccessRights.None would then not be a flag, right? It would be more like ReadWrite or All, not?

**Comment by Leogiciel** [comment]
@FredyFerrari considering [this MSDN article](https://msdn.microsoft.com/en-us/library/system.flagsattribute(v=vs.110).aspx), None seems to be a good practice as a default value, since All seems not (All is just the addition of all flags, 15 in my example, not 16).

I just think a flag keyword implementing best practices (None = 0, First=1, Second=2, Third=4, (...)) would save a lot of time to developpers with big enum flags, encapsulating the technical logic of the bitfield enum.

**Comment by FredyFerrari** [comment]
@Leogiciel so you have a maximum of 33 flags in a int32 based enum? None, Flag0, Flag1 .. Flag31? Sounds kind of wired, not? ;)

**Comment by jnm2** [comment]
Most of my flags enums do not have a `0` because people just type `0` when they mean 'no flags.' I dislike the arbitrariness of the sequence 0, 1, 2, 4, 8.

**Comment by Leogiciel** [comment]
Maybe because 0 is the default value ?

**Comment by jnm2** [comment]
@Leogiciel So the real question is, will people be able to commonly use your proposed sequence? Or will everyone be opting out all the time? To have value, the sequence needs to address an overwhelmingly common pattern.

**Comment by weitzhandler** [comment]
I'm fine with `[Flags]`.
